File: ai-vs-real-detector/api_client.ts
/**
 * TypeScript/JavaScript client for AI Detector Modal API
 * Use in Vercel, Next.js, or any Node.js/Browser environment
 */

export interface PredictionResult {
  label: string;
  score: number;
}

export interface PredictionResponse {
  predictions: PredictionResult[];
  top_prediction: string;
  confidence: number;
}

export interface BatchPredictionResponse {
  results: Array<PredictionResponse | { error: string }>;
}

export interface HealthCheckResponse {
  status: string;
  device: string;
  cuda_available: boolean;
  model_loaded: boolean;
}

export class AIDetectorClient {
  private apiUrl: string;
  private apiKey?: string;

  constructor(apiUrl: string, apiKey?: string) {
    this.apiUrl = apiUrl.replace(/\/$/, ''); // Remove trailing slash
    this.apiKey = apiKey;
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const headers: HeadersInit = {
      ...options.headers,
    };

    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const response = await fetch(`${this.apiUrl}${endpoint}`, {
      ...options,
      headers,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: 'Unknown error' }));
      throw new Error(error.error || `HTTP ${response.status}`);
    }

    return response.json();
  }

  /**
   * Convert File/Blob to base64 string
   */
  private async fileToBase64(file: File | Blob): Promise<string> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => {
        const base64 = (reader.result as string).split(',')[1];
        resolve(base64);
      };
      reader.onerror = reject;
      reader.readAsDataURL(file);
    });
  }

  /**
   * Predict from base64 encoded image
   */
  async predict(
    imageBase64: string,
    returnAllScores: boolean = true
  ): Promise<PredictionResponse> {
    return this.request<PredictionResponse>('/predict', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        image: imageBase64,
        return_all_scores: returnAllScores,
      }),
    });
  }

  /**
   * Predict from File object (browser)
   */
  async predictFile(
    file: File,
    returnAllScores: boolean = true
  ): Promise<PredictionResponse> {
    const base64 = await this.fileToBase64(file);
    return this.predict(base64, returnAllScores);
  }

  /**
   * Predict by uploading file
   */
  async predictUpload(file: File): Promise<PredictionResponse> {
    const formData = new FormData();
    formData.append('file', file);

    return this.request<PredictionResponse>('/predict/upload', {
      method: 'POST',
      body: formData,
    });
  }

  /**
   * Predict multiple images
   */
  async predictBatch(
    images: string[] | File[]
  ): Promise<BatchPredictionResponse> {
    let imageBase64Array: string[];

    if (images.length > 0 && images[0] instanceof File) {
      // Convert Files to base64
      imageBase64Array = await Promise.all(
        (images as File[]).map((file) => this.fileToBase64(file))
      );
    } else {
      imageBase64Array = images as string[];
    }

    return this.request<BatchPredictionResponse>('/predict/batch', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        images: imageBase64Array,
      }),
    });
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<HealthCheckResponse> {
    return this.request<HealthCheckResponse>('/health');
  }

  /**
   * Get API info
   */
  async getInfo(): Promise<any> {
    return this.request('/');
  }
}

// ============================================================================
// React Hooks (Optional)
// ============================================================================

/**
 * React hook for AI detection
 * 
 * Usage:
 *   const { predict, isLoading, result, error } = useAIDetector(apiUrl);
 *   await predict(file);
 */
export function useAIDetector(apiUrl: string, apiKey?: string) {
  const [isLoading, setIsLoading] = useState(false);
  const [result, setResult] = useState<PredictionResponse | null>(null);
  const [error, setError] = useState<string | null>(null);

  const client = useMemo(
    () => new AIDetectorClient(apiUrl, apiKey),
    [apiUrl, apiKey]
  );

  const predict = useCallback(
    async (file: File) => {
      setIsLoading(true);
      setError(null);

      try {
        const prediction = await client.predictFile(file);
        setResult(prediction);
        return prediction;
      } catch (err) {
        const errorMessage = err instanceof Error ? err.message : 'Prediction failed';
        setError(errorMessage);
        throw err;
      } finally {
        setIsLoading(false);
      }
    },
    [client]
  );

  const reset = useCallback(() => {
    setResult(null);
    setError(null);
  }, []);

  return {
    predict,
    isLoading,
    result,
    error,
    reset,
  };
}

// Add React imports for the hook
import { useState, useCallback, useMemo } from 'react';

// ============================================================================
// Next.js Server-Side Example
// ============================================================================

/**
 * Example Next.js API route (app/api/detect/route.ts)
 */
/*
import { NextRequest, NextResponse } from 'next/server';
import { AIDetectorClient } from '@/lib/ai-detector-client';

export async function POST(request: NextRequest) {
  try {
    const formData = await request.formData();
    const file = formData.get('file') as File;

    if (!file) {
      return NextResponse.json({ error: 'No file provided' }, { status: 400 });
    }

    const client = new AIDetectorClient(process.env.MODAL_API_URL!);
    const result = await client.predictUpload(file);

    return NextResponse.json(result);
  } catch (error) {
    console.error('Error:', error);
    return NextResponse.json(
      { error: 'Failed to process image' },
      { status: 500 }
    );
  }
}
*/

// ============================================================================
// Browser Example
// ============================================================================

/**
 * Example browser usage
 */
/*
const client = new AIDetectorClient('https://your-app.modal.run');

// Handle file upload
document.getElementById('fileInput').addEventListener('change', async (e) => {
  const file = e.target.files[0];
  
  try {
    const result = await client.predictFile(file);
    console.log('Prediction:', result.top_prediction);
    console.log('Confidence:', (result.confidence * 100).toFixed(1) + '%');
  } catch (error) {
    console.error('Error:', error);
  }
});
*/

export default AIDetectorClient;

File: ai-vs-real-detector/api_client.py
"""
Example client library for interacting with the Modal API
Can be used in Vercel, Supabase Edge Functions, or any Python application
"""
import requests
import base64
from typing import Dict, List, Optional, Union
from pathlib import Path
import io
from PIL import Image


class AIDetectorClient:
    """
    Client for the AI vs Real Image Detector Modal API
    
    Usage:
        client = AIDetectorClient("https://your-app.modal.run")
        result = client.predict_file("image.jpg")
        print(f"Prediction: {result['top_prediction']} ({result['confidence']*100}%)")
    """
    
    def __init__(self, api_url: str, api_key: Optional[str] = None):
        """
        Initialize the client
        
        Args:
            api_url: Base URL of your Modal API endpoint
            api_key: Optional API key for authentication
        """
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key
        self.headers = {}
        
        if api_key:
            self.headers["Authorization"] = f"Bearer {api_key}"
    
    def _encode_image(self, image: Union[str, Path, bytes, Image.Image]) -> str:
        """
        Encode image to base64 string
        
        Args:
            image: File path, bytes, or PIL Image
            
        Returns:
            Base64 encoded string
        """
        if isinstance(image, (str, Path)):
            # Read from file
            with open(image, "rb") as f:
                image_bytes = f.read()
        elif isinstance(image, bytes):
            image_bytes = image
        elif isinstance(image, Image.Image):
            # Convert PIL Image to bytes
            buffer = io.BytesIO()
            image.save(buffer, format="PNG")
            image_bytes = buffer.getvalue()
        else:
            raise ValueError(f"Unsupported image type: {type(image)}")
        
        return base64.b64encode(image_bytes).decode()
    
    def predict(
        self, 
        image_base64: str, 
        return_all_scores: bool = True
    ) -> Dict:
        """
        Predict from base64 encoded image
        
        Args:
            image_base64: Base64 encoded image string
            return_all_scores: Whether to return all class scores
            
        Returns:
            Prediction result dictionary
        """
        response = requests.post(
            f"{self.api_url}/predict",
            headers={**self.headers, "Content-Type": "application/json"},
            json={
                "image": image_base64,
                "return_all_scores": return_all_scores
            }
        )
        response.raise_for_status()
        return response.json()
    
    def predict_file(
        self, 
        image_path: Union[str, Path],
        return_all_scores: bool = True
    ) -> Dict:
        """
        Predict from image file
        
        Args:
            image_path: Path to image file
            return_all_scores: Whether to return all class scores
            
        Returns:
            Prediction result dictionary
        """
        image_base64 = self._encode_image(image_path)
        return self.predict(image_base64, return_all_scores)
    
    def predict_upload(self, image_path: Union[str, Path]) -> Dict:
        """
        Predict by uploading image file
        
        Args:
            image_path: Path to image file
            
        Returns:
            Prediction result dictionary
        """
        with open(image_path, "rb") as f:
            files = {"file": f}
            response = requests.post(
                f"{self.api_url}/predict/upload",
                headers=self.headers,
                files=files
            )
        response.raise_for_status()
        return response.json()
    
    def predict_batch(
        self, 
        images: List[Union[str, Path, bytes, Image.Image]]
    ) -> Dict:
        """
        Predict multiple images in one request
        
        Args:
            images: List of image paths, bytes, or PIL Images
            
        Returns:
            Batch prediction results
        """
        encoded_images = [self._encode_image(img) for img in images]
        
        response = requests.post(
            f"{self.api_url}/predict/batch",
            headers={**self.headers, "Content-Type": "application/json"},
            json={"images": encoded_images}
        )
        response.raise_for_status()
        return response.json()
    
    def health_check(self) -> Dict:
        """
        Check API health status
        
        Returns:
            Health status dictionary
        """
        response = requests.get(
            f"{self.api_url}/health",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def get_info(self) -> Dict:
        """
        Get API information
        
        Returns:
            API info dictionary
        """
        response = requests.get(
            f"{self.api_url}/",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()


# ============================================================================
# Example Usage
# ============================================================================

def example_single_prediction():
    """Example: Predict single image"""
    client = AIDetectorClient("https://your-app.modal.run")
    
    result = client.predict_file("test_image.jpg")
    
    print("📊 Prediction Results:")
    print(f"  Verdict: {result['top_prediction']}")
    print(f"  Confidence: {result['confidence']*100:.1f}%")
    print("\n  All scores:")
    for pred in result['predictions']:
        print(f"    {pred['label']}: {pred['score']*100:.2f}%")


def example_batch_prediction():
    """Example: Predict multiple images"""
    client = AIDetectorClient("https://your-app.modal.run")
    
    images = ["image1.jpg", "image2.jpg", "image3.jpg"]
    results = client.predict_batch(images)
    
    print("📊 Batch Prediction Results:")
    for i, result in enumerate(results['results']):
        print(f"\n  Image {i+1}:")
        print(f"    Verdict: {result['top_prediction']}")
        print(f"    Confidence: {result['confidence']*100:.1f}%")


def example_with_pil():
    """Example: Predict from PIL Image"""
    from PIL import Image
    
    client = AIDetectorClient("https://your-app.modal.run")
    
    # Open and process image
    image = Image.open("test_image.jpg")
    image = image.resize((224, 224))
    
    # Predict
    image_base64 = client._encode_image(image)
    result = client.predict(image_base64)
    
    print(f"Prediction: {result['top_prediction']}")


def example_health_check():
    """Example: Check API health"""
    client = AIDetectorClient("https://your-app.modal.run")
    
    health = client.health_check()
    print("🏥 Health Status:")
    print(f"  Status: {health['status']}")
    print(f"  Device: {health['device']}")
    print(f"  CUDA Available: {health['cuda_available']}")


if __name__ == "__main__":
    # Run examples
    print("Testing AI Detector API Client\n")
    
    try:
        example_health_check()
        print("\n" + "="*50 + "\n")
        example_single_prediction()
    except Exception as e:
        print(f"Error: {e}")
        print("\nMake sure to:")
        print("1. Update API URL in examples")
        print("2. Deploy your Modal app")
        print("3. Have test images available")
